source/azure_iot.c

```diff
@@ -0,0 +1,438 @@
+                                                   uint32_t ulValueLength )
+{
+    az_span xNameSpan = az_span_create( ( uint8_t * ) pucName, ( int32_t ) ulNameLength );
+    az_span xValueSpan = az_span_create( ( uint8_t * ) pucValue, ( int32_t ) ulValueLength );
+    az_result xResult;
+
+    if( pxMessageProperties == NULL )
+    {
+        AZLogError( ( "AzureIoTMessagePropertiesAppend failed: Invalid argument" ) );
+        return eAzureIoTInvalidArgument;
+    }
+
+    xResult = az_iot_message_properties_append( &pxMessageProperties->_internal.xProperties,
+                                                xNameSpan, xValueSpan );
+
+    if( az_result_failed( xResult ) )
+    {
+        return eAzureIoTFailed;
+    }
+
+    return eAzureIoTSuccess;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTResult_t AzureIoT_MessagePropertiesFind( AzureIoTMessageProperties_t * pxMessageProperties,
+                                                 const uint8_t * pucName,
+                                                 uint32_t ulNameLength,
+                                                 const uint8_t ** ppucOutValue,
+                                                 uint32_t * pulOutValueLength )
+{
+    az_span xNameSpan = az_span_create( ( uint8_t * ) pucName, ( int32_t ) ulNameLength );
+    az_span xOutValueSpan;
+    az_result xResult;
+
+    if( ( pxMessageProperties == NULL ) ||
+        ( ppucOutValue == NULL ) || ( pulOutValueLength == NULL ) )
+    {
+        AZLogError( ( "AzureIoTMessagePropertiesFind failed: Invalid argument" ) );
+        return eAzureIoTInvalidArgument;
+    }
+
+    xResult = az_iot_message_properties_find( &pxMessageProperties->_internal.xProperties,
+                                              xNameSpan, &xOutValueSpan );
+
+    if( az_result_failed( xResult ) )
+    {
+        return eAzureIoTItemNotFound;
+    }
+
+    *ppucOutValue = az_span_ptr( xOutValueSpan );
+    *pulOutValueLength = ( uint32_t ) az_span_size( xOutValueSpan );
+
+    return eAzureIoTSuccess;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTResult_t AzureIoT_Base64HMACCalculate( AzureIoTGetHMACFunc_t xAzureIoTHMACFunction,
+                                               const uint8_t * pucKey,
+                                               uint32_t ulKeySize,
+                                               const uint8_t * pucMessage,
+                                               uint32_t ulMessageSize,
+                                               uint8_t * pucBuffer,
+                                               uint32_t ulBufferLength,
+                                               uint8_t * pucOutput,
+                                               uint32_t ulOutputSize,
+                                               uint32_t * pulOutputLength )
+{
+    AzureIoTResult_t xStatus;
+    uint8_t * pucHashBuf;
```

Comment:
Nit - consider a new variable for `uint8_t * pucDecodedKeyBuf = pucBuffer;` and then use `pucDecodedKeyBuf` instead of `pucBuffer` when calculating and using the key.

One the one hand I hesitate on this because it's aliasing.  But on other hand you have a similar pattern with `pucHashBuf` pointing to the place in the passed user buffer where it needs to write to.  

I suggest above because the intent is much clearer with the name to not just be buffer.

Reply:
fair. makes readability easier